src/EncryptedModels.ts

```diff
@@ -309,31 +309,34 @@ class EncryptedRevision<CM extends CollectionItemCryptoManager> {
       chunks.push([hash, buf]);
     }
 
-    // Shuffle the items and save the ordering if we have more than one
-    if (chunks.length > 0) {
+    // Shuffle the items, deduplicate and save the ordering
+    if (chunks.length > 1) {
       const indices = shuffle(chunks);
+      const uniqueChunksMap = new Map<string, [base64, Uint8Array]>();
 
-      // Filter duplicates and construct the indice list.
-      const uidIndices = new Map<string, number>();
-      chunks = chunks.filter((chunk, i) => {
+      chunks.forEach((chunk) => {
         const uid = chunk[0];
-        const previousIndex = uidIndices.get(uid);
-        if (previousIndex !== undefined) {
-          indices[i] = previousIndex;
-          return false;
-        } else {
-          uidIndices.set(uid, i);
-          return true;
+        if (!uniqueChunksMap.has(uid)) {
+          uniqueChunksMap.set(uid, chunk);
         }
       });
 
-      // If we have more than one chunk we need to encode the mapping header in the last chunk
-      if (indices.length > 1) {
-        // We encode it in an array so we can extend it later on if needed
-        const buf = msgpackEncode([indices]);
-        const hash = toBase64(cryptoManager.calculateMac(buf));
-        chunks.push([hash, buf]);
-      }
+      const chunkKeys = [...uniqueChunksMap.keys()];
+
+      // Change the original (shuffled) indices to point at the deduplicated chunks
+      const newIndices = indices.map((i) => {
+        const [id] = chunks[i];
+        return chunkKeys.indexOf(id);
+      });
+
+      chunks = [...uniqueChunksMap.values()];
```

Comment:
Sorry it took me so long to review this. It's a very intrusive change in a very sensitive place.

I think the change is overall correct, though one thing that I'm very concerned about is this highlighted code I'm responding to. I'm not sure that per spec keys and values are necessarily in the same order and that it's deterministic. It could very well be, though I'm unsure. I'd much rather we created both in one go.

I'm also not sure what uniqueChunksMap actually does to the ordering.

I'll revert back to Rust, this is how it's done there, do you think that makes sense?
```rust
            // Filter duplicates and construct the indice list.
            let mut uid_indices: HashMap<String, usize> = HashMap::new();
            chunks = chunks
                .into_iter()
                .enumerate()
                .filter_map(|(i, chunk)| {
                    let uid = &chunk.0;
                    match uid_indices.get(uid) {
                        Some(previous_index) => {
                            indices[i] = *previous_index;
                            None
                        }
                        None => {
                            uid_indices.insert(uid.to_string(), i);
                            Some(chunk)
                        }
                    }
                })
                .collect();

            // If we have more than one chunk we need to encode the mapping header in the last chunk
            if indices.len() > 1 {
                // We encode it in an array so we can extend it later on if needed
                let buf = rmp_serde::to_vec_named(&(indices,))?;
                let hash = to_base64(&crypto_manager.0.calculate_mac(&buf)?)?;
                chunks.push(ChunkArrayItem(hash, Some(buf)));
            }
```